src/disco/gui/fd_gui.c

```diff
@@ -1618,96 +1620,125 @@ fd_gui_plugin_message( fd_gui_t *    gui,
   fd_gui_slot_t * slot = gui->slots[ _slot % FD_GUI_SLOTS_CNT ];
-  if( FD_UNLIKELY( slot->slot!=_slot ) ) fd_gui_clear_slot( gui, _slot, ULONG_MAX );
-
-  slot->cus.leader_start_time  = start_time_nanos;
 
-  long tickcount = fd_tickcount();
-  if( FD_LIKELY( slot->cus.reference_ticks==LONG_MAX ) ) slot->cus.reference_ticks = tickcount;
-  slot->cus.reference_nanos = fd_log_wallclock() - (long)((double)(tickcount - slot->cus.reference_ticks) / fd_tempo_tick_per_ns( NULL ));
-
-  slot->cus.leader_end_time   = end_time_nanos;
-  slot->cus.max_compute_units = (uint)max_compute_units;
-  if( FD_LIKELY( slot->cus.microblocks_upper_bound==USHORT_MAX ) ) slot->cus.microblocks_upper_bound = (ushort)max_microblocks;
+  slot->txs.leader_start_time = start_time_nanos;
+  slot->txs.leader_end_time   = end_time_nanos;
+  slot->txs.max_compute_units = (uint)max_compute_units;
+  if( FD_LIKELY( slot->txs.microblocks_upper_bound==USHORT_MAX ) ) slot->txs.microblocks_upper_bound = (ushort)max_microblocks;
 }
 
 void
 fd_gui_unbecame_leader( fd_gui_t * gui,
+                        long       tickcount,
                         ulong      _slot,
                         ulong      microblocks_in_slot ) {
+  fd_gui_init_slot_txns( gui, tickcount, _slot );
   fd_gui_slot_t * slot = gui->slots[ _slot % FD_GUI_SLOTS_CNT ];
-  if( FD_UNLIKELY( slot->slot!=_slot ) ) fd_gui_clear_slot( gui, _slot, ULONG_MAX );
 
-  slot->cus.microblocks_upper_bound = (ushort)microblocks_in_slot;
+  slot->txs.microblocks_upper_bound = (ushort)microblocks_in_slot;
 }
 
 void
-fd_gui_execution_begin( fd_gui_t *   gui,
-                        long         tickcount,
-                        ulong        _slot,
-                        ulong        bank_idx,
-                        ulong        txn_cnt,
-                        fd_txn_p_t * txns ) {
+fd_gui_microblock_execution_begin( fd_gui_t *   gui,
+                                   long         tickcount,
+                                   ulong        _slot,
+                                   fd_txn_p_t * txns,
+                                   ulong        txn_cnt,
+                                   ulong        pack_txn_idx,
+                                   int          is_bundle ) {
+  fd_gui_init_slot_txns( gui, tickcount, _slot );
   fd_gui_slot_t * slot = gui->slots[ _slot % FD_GUI_SLOTS_CNT ];
-  if( FD_UNLIKELY( slot->slot!=_slot ) ) fd_gui_clear_slot( gui, _slot, ULONG_MAX );
 
-  if( FD_UNLIKELY( !slot->cus.has_offset[ 64UL ] ) ) slot->cus.start_offset[ 64UL ] = gui->cus.offset;
-  slot->cus.has_offset[ 64UL ] = 1;
-  if( FD_UNLIKELY( slot->cus.reference_ticks==LONG_MAX ) ) slot->cus.reference_ticks = tickcount;
+  if( FD_UNLIKELY( slot->txs.start_offset==ULONG_MAX ) ) slot->txs.start_offset = pack_txn_idx;
+  else                                                   slot->txs.start_offset = fd_ulong_min( slot->txs.start_offset, pack_txn_idx );
+
+  gui->pack_txn_idx = fd_ulong_max( gui->pack_txn_idx, pack_txn_idx );
```

Comment:
You have a strong invariant here that must be true, so I'd assert it

```suggestion
  FD_TEST( pack_txn_idx==gui->pack_txn_idx+1UL );
  gui->pack_txn_idx++;
```

Reply:
I though stem could reorder frags on a link? If the order of two microblocks gets inverted wouldn't this "break" the invariant?

Also, `pack_txn_idx` might increment by more than 1 for bundles. Since this variable is used to detect if a past slot is overwritten I should probably be doing

```
gui->pack_txn_idx = fd_ulong_max( gui->pack_txn_idx, pack_txn_idx+txn_cnt-1UL )
```